plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java

```diff
@@ -1342,6 +1507,71 @@ public boolean upgradeKubernetesCluster(UpgradeKubernetesClusterCmd cmd) throws
         return upgradeWorker.upgradeCluster();
     }
 
+    @Override
+    public boolean addVmsToCluster(AddVirtualMachinesToKubernetesClusterCmd cmd) {
```

Comment:
I see no validation code for the actual VM, what are we trusting to the user and what is automagically done by the system?
I understand that we require due diligence from the user. No issue there. We should be able to document that properly though.

Reply:
What kind of validation do you mean here? As per my understanding, the below code takes care of the VM validation (that the uuid of the VM exists). And later on we also check if the user has access to the VM or not.

```
    @Parameter(name = ApiConstants.VIRTUAL_MACHINE_IDS, type = CommandType.LIST,
            collectionType=CommandType.UUID,
            entityType = UserVmResponse.class,
            required = true,
            description = "the IDs of the VMs to add to the cluster")
```

Reply:
I mean about the contents of the VM and its validity to perform its role in/for the cluster.

It should be provisioned with kubernetes code and maybe even containers, should it?

Reply:
The feature is just to list down VMs which were created by an external kubernetes provider on CloudStack.
I don't think we can we verify what a user is running within a VM.

Reply:
i understand (though i think some checks are possible) but this must be well documented.
I wonder also if we can do diagnosis on any clusters that were imported. Why do they fail and what can be done. i might be a hypochondriac but I foresee a lot of user questions on this feature ;)